### PR TITLE
react-dom(experimental): Add `preload()` and `preinit()`

### PR DESCRIPTION
**Pull request by eps1lon**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
   - [`preload` implementation](https://github.com/facebook/react/blob/9cdf8a99edcfd94d7420835ea663edca04237527/packages/react-dom-bindings/src/client/ReactDOMFloatClient.js#L164-L166)
   - [`preinit` implementation](https://github.com/facebook/react/blob/9cdf8a99edcfd94d7420835ea663edca04237527/packages/react-dom-bindings/src/client/ReactDOMFloatClient.js#L217-L224)
- ~[ ]~ If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

/cc @sebmarkbage @gnoff

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-dom ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-dom@latest/))
was missing the following properties:
1. createRoot
1. hydrateRoot
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against cd04b8c802e398448d8e39271eb76306c9620209
</p>

**Review by andrewbranch — APPROVED** [review]

File: types/react-dom/experimental.d.ts [hunk 1]
```diff
@@ -32,4 +32,21 @@ import ReactDOM = require('./next');
 
 export {};
 
-declare module '.' {}
+declare module '.' {
```
Comment (andrewbranch) [inline types/react-dom/experimental.d.ts:35]: What is this trick doing? Extensionless module specifier in ESM?

Comment (eps1lon) [reply types/react-dom/experimental.d.ts:35]: It's referencing the current folder which then falls back to `index`. Could just write `react-dom` here but I wanted to be like the cool kids, write code that I can simply c&p into other modules (React has the same pattern) thus making my code DRY 😎

**Comment by eps1lon** [comment]
Superseded by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62849

**Pull request closed without merging**